src/box/relay.cc

```diff
@@ -1119,6 +1119,14 @@ static void
 relay_send(struct relay *relay, struct xrow_header *packet)
 {
 	ERROR_INJECT_YIELD(ERRINJ_RELAY_SEND_DELAY);
+	/**
+	 * Until Tarantool 3.2.1 all raft requests were sent with
+	 * GROUP_LOCAL id. In order not to break the upgrade process,
+	 * we still send local rows to old replicas.
+	 */
+	if (iproto_type_is_raft_request(packet->type) &&
+	    relay->version_id < version_id(3, 2, 1))
+		packet->group_id = GROUP_LOCAL;
```

Comment:
Hm. Having it in such a generic place feels quite crutchy, like you said. Perhaps extract Raft row sending into a new function called `relay_send_raft()` containing this check and invoke it explicitly in `relay_send_tx()` (when request type is Raft) and in `relay_raft_msg_push()` unconditionally?

Then we could at least isolate the hack.

Reply:
WDYT of the current implementation? `relay_send_tx` never sends raft, all local rows are filtered out from it. You meant `relay_send_initial_join_row`, I suppose

Reply:
Aah, yep, then in `relay_send_initial_join_row()`. The rest of the implementation is fine! I can't think of any other way.

Reply:
@Gerold103, please, approve to PR, if we're good to go)